### PR TITLE
8269119: C2: Avoid redundant memory barriers in Unsafe.copyMemory0 intrinsic

### PR DESCRIPTION
**Pull request by iwanowww**
`Unsafe::copyMemory0` intrinsic unconditionally inserts barriers arounds the call to `unsafe_arraycopy` stub.
It is a conservative approach and barriers can be avoided in most common cases (similar to what is done for scalar unsafe accesses).

`Unsafe::copyMemory()` performs argument validation which limits inputs either
to off-heap location (null + absolute address) or primitive on-heap array.

The only cases when barriers are still needed are:
 * mixed accesses (`Object+offset`);
 * mismatched access due to lack of type info on base oop (`Object:NotNull+offset`). 

Testing: hs-tier1 - hs-tier6

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269119](https://bugs.openjdk.java.net/browse/JDK-8269119): C2: Avoid redundant memory barriers in Unsafe.copyMemory0 intrinsic

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Roland Westrelin](https://openjdk.java.net/census#roland) (@rwestrel - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5259/head:pull/5259` \
`$ git checkout pull/5259`

Update a local copy of the PR: \
`$ git checkout pull/5259` \
`$ git pull https://git.openjdk.java.net/jdk pull/5259/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5259`

View PR using the GUI difftool: \
`$ git pr show -t 5259`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5259.diff">https://git.openjdk.java.net/jdk/pull/5259.diff</a>

</details>

### CODE REVIEW

**Comment by iwanowww** [comment]
Good point, Roland. 

Removed barrier-related logic and use `TypePtr::BOTTOM` instead.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: src/hotspot/share/opto/library_call.cpp [hunk 1]
```diff
@@ -4070,6 +4070,18 @@ bool LibraryCallKit::inline_fp_conversions(vmIntrinsics::ID id) {
...
+//----------------------has_wide_mem-------------------------
+bool LibraryCallKit::has_wide_mem(Node* addr, Node* base) {
+  const Type* base_t = _gvn.type(base);
+
+  bool in_native = (base_t == TypePtr::NULL_PTR);
+  bool in_heap   = !TypePtr::NULL_PTR->higher_equal(base_t);
+  bool is_mixed  = !in_heap && !in_native;
+
+  bool is_array = _gvn.type(addr)->isa_aryptr();
+  return is_mixed || (in_heap && !is_array);
```
Comment (rwestrel) [inline src/hotspot/share/opto/library_call.cpp:4082]: Why is the is_array check needed?

Comment (rwestrel) [reply src/hotspot/share/opto/library_call.cpp:4082]: Tobias explained that one to me privately

File: src/hotspot/share/opto/library_call.cpp [hunk 2]
```diff
@@ -4070,6 +4070,18 @@ bool LibraryCallKit::inline_fp_conversions(vmIntrinsics::ID id) {
   return true;
 }
 
+//----------------------has_wide_mem-------------------------
+bool LibraryCallKit::has_wide_mem(Node* addr, Node* base) {
```
Comment (rwestrel) [inline src/hotspot/share/opto/library_call.cpp:4074]: Why do you need both arguments? Isn't addr sufficient?

**Review by rwestrel — APPROVED** [review]
Looks good to me.

**Comment by iwanowww** [comment]
Thinking more about it, I spotted that `is_array` check is not enough: it doesn't cover `bottom[]` case which should be treated as wide memory. Fixed in the latest version.

File: test/hotspot/jtreg/compiler/unsafe/UnsafeCopyMemory.java [hunk 3]
```diff
@@ -28,6 +28,7 @@
  * @modules java.base/jdk.internal.misc
  *
  * @run main/othervm -Xbatch -XX:CompileCommand=dontinline,compiler.unsafe.UnsafeCopyMemory::test*
+ *                   -XX:+UnlockDiagnosticVMOptions -XX:+StressGCM -XX:+StressLCM
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/unsafe/UnsafeCopyMemory.java:31]: You should add `@key stress` to the test.

File: src/hotspot/share/opto/library_call.cpp [hunk 4]
```diff
@@ -4070,6 +4070,20 @@ bool LibraryCallKit::inline_fp_conversions(vmIntrinsics::ID id) {
...
+  const TypeAryPtr* addr_t = _gvn.type(addr)->isa_aryptr();
+  const Type*       base_t = _gvn.type(base);
+
+  bool in_native = (base_t == TypePtr::NULL_PTR);
+  bool in_heap   = !TypePtr::NULL_PTR->higher_equal(base_t);
+  bool is_mixed  = !in_heap && !in_native;
+
+  bool is_prim_array = (addr_t != NULL) && (addr_t->elem() != Type::BOTTOM);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/library_call.cpp:4082]: What does `is_prim_array` stand for? I read this as "is primitive array" which does not make sense :)

Comment (iwanowww) [reply src/hotspot/share/opto/library_call.cpp]: Yes, it stands for "primitive array". 
Unsafe.copyMemory() validates inputs using Unsafe.checkPrimitiveArray() to ensure they are primitive arrays for on-heap accesses. I refactored `has_wide_mem` a bit to make it a bit clearer what's going on.

File: test/hotspot/jtreg/compiler/unsafe/UnsafeCopyMemory.java (hunk 3 above)
Comment (iwanowww) [reply test/hotspot/jtreg/compiler/unsafe/UnsafeCopyMemory.java]: Fixed.

**Review by TobiHartmann — APPROVED** [review]
Thanks for the details, looks good to me.

**Review by rwestrel — APPROVED** [review]

**Comment by iwanowww** [comment]
Thanks for the reviews, Tobias and Roland.

/integrate

**Pull request closed without merging**